File: learning/exceptions.py
# exceptions
#   - Exception Handling
#       - Python has five main ways of exception handling
#           - try
#               - runs a block of code and "tests" for errors or exceptions while being executed
#               - any errors will be intercepted by the except statement
#               - syntax:
#                   try:
#                       <stataments>
#               - example:
#                   try:
#                       print("lumbago")
#           - except
#               - runs a block of code after the Exception is run
#               - can catch any exceptions in a program
#                   - from broad to specific Exceptions
#               - requires a try statement before it
#               - syntax:
#                   except <exception>:
#                       <statament>
#               - example:
#                   except:             #> catches all exceptions
#                       print('error')
#           - finally
#               - runs regardless of the result of the try-except stataments
#               - runs after a try-except
#               - requires a try-except statement before it
#               - syntax:
#                   finally:
#                       <statement>
#               - example:
#                   finally:
#                       print('code finished')
#           - else
#               - runs only if the try-except statements returns no errors
#               - requires a try-except statememt before it
#               - syntax:
#                   else:
#                       <statements>
#               - example:
#                   else:
#                       print('no errors')
#           - raise
#               - used to manually throw or raise exceptions
#               - typically used in a try-except statement
#                   - it can be used outside but note that the program will stop after an exception is thrown
#               - syntax:
#                   raise <exception>
#               - example:
#                   raise ArithmeticError
#
#       - multiple except blocks are able to declared in a try-except statement
#           - it can help in handling multiple exceptions by being more specific
#           - example:
#               try:
#                   my_list = [1, 2, 3, 4]
#                   my_list[5]             #> IndexError
#                   my_list[1] / 0         #> ZeroDivisionError
#               except IndexError as err:
#                   print('index out of bounds yo')
#               except ZeroDivisionError as err:
#                   print('idiot, cannot by zero bro')
#               finally:
#                   print('program finished')
#               else:
#                   print('program finished without errors, 5 starts')

File: java/src/learning/Exceptions.java
/* exceptions
    - Java Errors
        - when Java encounters an error, it immediately stops the program
        - unlike Exceptions, it cannot be circumvented unless it's fixed
        - three types of errors
            - Compile-Time Errors
                - occurs duing compilation time
                    - preventing the file from compiling properly
                    - mainly due to syntax and type errors
                - common examples are:
                    - missing semilcolons
                        int x = 5
                    - incompatible types
                        boolean yes = "Hello";
                    - undeclared variables
                        System.out.print(myValue);
            - Runtime Errors
                - occurs after the program compiles
                    - but encounters sudden crashes or unintended behaviors
                - common examples are:
                    - division by zero
                        int x = 10 / 0;
                    - index out of bounds
                        String arr[] = new String[1];
                        arr[10] = "yes";
            - Logical Errors
                - occurs no Compilation Errors or Runtime Errors occur
                    - the output is not accurate to what was supposed to happen
                - common examples are
                    - expecting a sum but the results is a difference
                        int x = 10;
                        int y = 8;
                        System.out.print("Sum: ", x - y);

        - how to avoid errors?
            - use meaningful variable names
                - avoid typos and spellings
            - check for missing semicolons and braces
            - read the error messages carefully
                - gives valuable information in debugging

        - debugging
            - the process of identifying and fixing errors or bugs present in a program
            - it often includes:
                - reading error messages
                - tracing program flow step by step
                - testing small pieces of code independently
            - common debugging steps:
                - printing values for each step
                    - mainly used when encountering Logical Errors
                    - can reveal which part of the program is failing
                    - example:
                        int x = 10;
                        System.out.print(x);
                        int y = 8;
                        System.out.print(y);
                - using IDEs for debugging
                    - modern Java IDEs comes with debugging tools
                        - examples: IntelliJ IDEA, Eclipse, Netbeans, etc.
                        - these are able to:
                            - add breakpoints
                                - dictates where the program ends for testing
                            - inspect variables in real time
                                - being able to see their values step by step
                - commenting 
                    - using the comment // syntax to rule out problematic lines of code
                    - example:
                        int x = 10;
                        // float y = "Hello"; -> Error
                        String yes = "no";
                - looking for typos
                    - calling for elements can sometimes yield in mispellings ot typos
                    - example:
                        int myValue = 10;
                        System.out.print(MyValue); // typo

    - Java Exception
        - different types of problems and errors can occur while a program is running
            - yielding different error messages depending on the type
        - when the program encounters a roadblock or an error, Java immediately stops the program
            - it then throws an error describing the reason for stopping the program
        - unlike Errors, these types of errors are able to be circumvented
            - commonly called as Exceptions 
        - these types of errors are "predicted" and have code that runs when the program runs into that error

    - Exception handling
        - Java has four keywords for Exception Handling
            - try
                - runs a block of code and "tests" it while being executed
                - any error messages will be "caught" by the catch statement
                - syntax:
                    try {...}
                - example:
                    try {
                        int x = 100;
                        int y = 0;
                        System.out.print(x / y); -> Error
                    }
            - catch
                - runs a block of code after the Exception it is "expecting" is caught
                - can catch any Exceptions, from broad to specific
                - sidenote:
                    - the catch statement can also catch Error, but is genereally unadvised
                        - Errors are a underlying problem with the program
                        - it is better to solve it to prevent futher problems
                - syntax:
                    catch(<exception> <identifier>) {...}
                - example:
                    catch(Exception e) {
                        System.out.print("Error encountered"); // prints when an Exception is triggered
                    }
            - finally 
                - runs regardless of the result of a try-catch statement
                - runs after a try-catch
                - syntax:
                    finally {...}
                - example:
                    finally {
                        System.out.print("try-catch is finished");
                    }
            - throw
                - used to throw Exceptions
                    - can also throw Errors but is ill-adviced
                - typically used together with an exception type
                - can be used outside a try-catch statement
                    - note that the program will be stopped after an Exception is thrown
                - syntax:
                    throw <objectName> <exceptionClass>;
                - example:
                    throw new Exception;

        - the try and catch keywords usually come in as pairs
            - with the finally statement being optional
            - syntax:
                try {...}
                catch(<exeption> <identifier>) {...}
                // optional
                finally {...}
            - example:
                try {
                    int i = 10 / 0;                       // triggers the ArithmeticException error
                } catch (ArithmethicException e) {
                    System.out.print(e);                  // prints after the Exception is triggered
                } finally {
                    System.out.print("Program finished"); // runs after the try-catch statement is done
                }

        - multiple catch blocks are able to be declared in a try-catch statement
            - it can help in handling multiple Exceptions through custom statements
            - note that catch order matters
                - exceptins closer to the try statement runs first when triggered
                - it is by convention to declare exceptions from specific to general Exceptions
            - example:
                    try {
                    int arr[] = {1, 2, 3};
                    arr[10] = 10;                                       // ArrayIndexOutOfBoundException
                    int yes = 10 / 0;                                   // ArithmeticException
                } catch (ArithmeticException e) {
                    System.out.print("arithmethic error"); 
                } catch (ArrayIndexOutOfBoundException e) {
                    System.out.print("arrayindexourofbound error"); 
                } catch (Exception e) {
                    System.out.print("broad exception");            // runs when previous exceptions are not triggered
                }
            - it is also possible to catch multiple exceptions in a single statement
                - it is through using the or | operator
                - it saves space by avoiding repetition
                - syntax:
                    catch(<exception1> | <exception2> e) {...}
                - example:
                    catch (ArithmeticException | ArrayIndexOutOfBoundsException e) {...}
*/

public class Exceptions {

}
